### PR TITLE
Add README for pluginrpc-gen

### LINKED ISSUE #16248: Document how to generate pluginrpc clases
As far as I know, we don't have documentation about how the volume, networking and other external driver proxies are generated. I don't think we have a make command to generate them. We should add documentation and a script to facilitate those tasks.

### PR DESCRIPTION
**Pull request by cpuguy83**
Closes #16248

ping @calavera

### CODE REVIEW

**Comment by calavera** [comment]
Nice, thank you! :sparkles: :rainbow: 

LGTM

**Comment by tiborvass** [comment]
LGTM

File: pkg/plugins/pluginrpc-gen/README.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
+Plugin RPC Generator
+====================
+
+Generates go code from a go interface definition for proxying between the plugin
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: Just a nit; s/go/Go/

File: pkg/plugins/pluginrpc-gen/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+type volumeDriver interface {
+	Create(name string, opts opts) (err error)
+	Remove(name string) (err error)
+	Path(name string) (mountpoint string, err error)
+	Mount(name string) (mountpoint string, err error)
+	Unmount(name string) (err error)
+}
+```
+
+**Note** All function options and return values must be named in the definition.
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: For notes, we usually do;

```Markdown
> **Note**: foo
> bar
```

File: pkg/plugins/pluginrpc-gen/README.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+Run the generator:
+
+```bash
+$ pluginrpc-gen -type volumeDriver -name VolumeDriver -i volumes/drivers/extpoint.go -o volumes/drivers/proxy.go
+```
+
+Where:
+- `-type` is the name of the interface to use
+- `-name` is the subsystem that the plugin "Implements"
+- `-i` is the input file containing the interface defintiion
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: typo s/defintiion/definition/

File: pkg/plugins/pluginrpc-gen/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+- `-o` is the output file where the the generated code should go
+
+**Note** The generated code will use the same package name as the one defined in the input file
+
+Optionally, you can skip functions on the interface should should not be
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: s/should should not/that should not/

File: pkg/plugins/pluginrpc-gen/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+implemented in the generated proxy code by passing in the function name to `-skip`.
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: Just something I noticed; Does this use single dashes for long options? (`-skip`, not `--skip`); would be more consistent with "Docker" to use double dashes.

File: pkg/plugins/pluginrpc-gen/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+type opts map[string]string
+```
+
+This `opts` type will work, using a `map[string]string` directly will not.
+
+## go-generate
+
+You can also use this with go-generate, which is pretty awesome.  
+To do so, place the code at the top of the file which contains the interface
+definition (ie, the input file):
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: s/ie/i.e./

File: pkg/plugins/pluginrpc-gen/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+
+```go
+//go:generate pluginrpc-gen -i $GOFILE -o proxy.go -type volumeDriver -name VolumeDriver
+```
+
+Then cd to the package dir and run `go generate`
+
+**Note** the `pluginrpc-gen` binary must be within your $PATH
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: $PATH -> `$PATH`

File: pkg/plugins/pluginrpc-gen/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+The parser can currently only handle types which are not specifically a map or
+a slice.  
+You can, however, create a type that uses a map or a slice internally, for instance:
+
+```go
+type opts map[string]string
+```
+
+This `opts` type will work, using a `map[string]string` directly will not.
```
Comment (thaJeztah) [inline pkg/plugins/pluginrpc-gen/README.md]: Perhaps; "whereas using a..."?

**Comment by thaJeztah** [comment]
Some nits, typos and a question, but roughly LGTM, thanks!

**Comment by cpuguy83** [comment]
Thanks @thaJeztah, those typos never seem to stop!

Updated.

**Comment by thaJeztah** [comment]
thanks; LGTM

**Pull request merged**